src/dmd/mtype.d

```diff
@@ -7278,3 +7278,37 @@ bool isCopyable(Type t)
     }
     return true;
 }
+
+/** Returns: `true` iff `t` has any non-`immutable` (mutable or const) indirections.
+ *
+ * In other words, if other references to `t` may be able to change it.
+ */
+bool hasAliasing(scope const Type t) pure nothrow @nogc
+{
+    if (t.isTypeEnum())
+        return false;           // an enum is an r-value so cannot alias
+    if (t.isImmutable)          // immutable is transitive
+        return false;           // eager bail out
+    if (const(TypeClass) tc = t.isTypeClass)
+    {
+        const(ClassDeclaration) cd = tc.sym;
+        foreach (const(VarDeclaration) vd; cd.fields)
+            if (hasAliasing(vd.type))
+                return true;
+    }
+    else if (const(TypeStruct) ts = t.isTypeStruct)
```

Comment:
That wastes an extra duplicate checking for `t` being an enum already checked at the top of `hasAliasing`.

Reply:
Hmm, when thinking about it I'm unsure how treat enum types. Have to check how `std.traits.hasAliasing` behaves for enum types.

Reply:
Furthermore, using `isAggregate` we loose both `pure nothrow @nogc` aswell as `const` qualifier on `Type`.

Reply:
Sorry, missread this function to work with the actual declaration instead of types.

You can merge class/struct anyways by extracting the sym from `TypeClass`/`TypeStruct` using the common type `Aggregate`. No need to copy & paste for both cases.

Reply:
But then again if we want to support correct behaviour of enum types maybe we need to use `toBaseType` anyhow. I'm not sure why that is not (yet) qualified.

Reply:
> Sorry, missread this function to work with the actual declaration instead of types.
> 
> You can merge class/struct anyways by extracting the sym from `TypeClass`/`TypeStruct` using the common type `Aggregate`. No need to copy & paste for both cases.

I just realized class fields shall not be recursed. Only struct fields.

Reply:
> Hmm, when thinking about it I'm unsure how treat enum types. Have to check how `std.traits.hasAliasing` behaves for enum types.

Support for enum types has been added to the top of `hasAliasing`.

Reply:
Marking this a solved.